docs/decisions/infra/0005-database-module-design.md

```diff
@@ -0,0 +1,63 @@
+* Option B: Separate the database infrastructure into a separate layer
+
+### Decision Outcome: Separate the database infrastructure into a separate layer
+
+Changes to database infrastructure are infrequent and therefore do not need to be incorporated as part of the continuous delivery process of deploying the application as it would needlessly slow down application deploys and also increase the risk of accidental changes to the database layer. When database changes are needed, they are sometimes complex due to the stateful nature of databases and can require multiple steps to make those changes gracefully. For these changes, it is beneficial to separate them from application resources so that application deploys can remain unaffected. Finally, breaking down the environment setup process into smaller, more linear steps – creating the database first before creating the application service – makes the environment setup process easier to understand and troubleshoot than trying to do create everything at once.
+
+The biggest disadvantage to this approach is the fact that dependencies between root modules cannot be directly expressed in terraform. To mitigate this problem, we should carefully design the interface between root modules to minimize breaking changes in that interface.
+
+## Pros and Cons of the Options
+
+### Option A: Put the database infrastructure in the same root module as the application service
+
+Pros:
+
+* This is what we've typically done in the past. All the infrastructure necessary for the application environment would live in a single root module, with the exception of shared resources like the ECR image repository.
+
+Cons:
+
+* The application service's healthcheck depends on the database cluster to be created and the database user to be provisioned. This cannot easily be done in a single terraform apply.
+* Changes to the database infrastructure are often more complex than changes to application infrastructure. Unlike application infrastructure, database changes cannot take the approach of spinning up new infrastructure in desired configuration, redirecting traffic to new infrastructure, then destroying old infrastructure. This is because application infrastructure can be designed to be stateless while databases are inherently stateful. In such cases, making database changes may require careful coordination and block changes to the application infrastructure, potentially including blocking deploys, while the database changes are made.
+* Application deploys would be slower since there is more state for terraform to fetch during the terraform plan and apply steps.
+
+### Option B: Separate the database infrastructure into a separate layer
+
+Pros:
+
+* Separating the database layer makes explicit the dependency between the database and the application service, and enables an environment setup process that involves only creating resources when all dependencies have been created first.
```

Comment:
Nice. I think something somewhat implied by this decision is basically treating the database as a separate application/service that should be deployed and updated separately (e.g. version upgrades, migrations, etc.), which then makes it a little more obvious that developers should be careful about breaking changes between the app and db.

Reply:
yeah. is this a side comment or are you asking to include something like this point explicitly in the list of Pros?

Reply:
just a side comment!